src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs

```diff
@@ -0,0 +1,428 @@
+
+            // First of all, we need to verify the source addresses to belong to the Merchant.
+            var sourcesCheckupResult = await CheckupSourcesAndWalletsAsync(transferRequestUpdated);
+            if (sourcesCheckupResult != null)
+                return sourcesCheckupResult;
+
+            // Here would be no additional validations. If transaction fails, we will know it from the underlying BitcoinApi service.
+
+            // Save the initial state of transfer
+            await _transferRepository.SaveAsync(transferRequestUpdated);
+            await _transferRequestPublisher.PublishAsync(transferRequestUpdated);
+
+            // The main work
+            foreach (var tran in transferRequestUpdated.TransactionRequests)
+            {
+                var sourceAddresses = (from s in tran.SourceAmounts
+                                       select new ToOneAddress(s.Address, s.Amount)).ToList();
+                // Please, note: feeRate = 0 and fixedFee = 0 here
+                var currentResult = await _bitcointApiClient.TransactionMultipleTransfer(null, tran.DestinationAddress, tran.Currency, 0, 0, sourceAddresses);
+
+                if (currentResult == null || currentResult.HasError)
+                {
+                    // Please, note: after the first faulty transaction we abort the further execution.
+                    // BUT money that we had already transfered will not go back to source address(es).
+                    // To make the whole transfer request atomic we need to implement such feature in
+                    // BitcoinApi service at first.
+                    return await PackFaultyResult(currentResult, tran, transferRequestUpdated);
+                }
+                else
+                {
+                    tran.TransactionHash = currentResult.Transaction.Hash;
+                }
+
+                await _transferRepository.SaveAsync(transferRequestUpdated);
+                await _transferRequestPublisher.PublishAsync(transferRequestUpdated);
+            }
+
+            transferRequestUpdated.TransferStatus = TransferStatus.InProgress;
+            transferRequestUpdated.TransferStatusError = TransferStatusError.NotError;
+            await _transferRepository.SaveAsync(transferRequestUpdated);
+            await _transferRequestPublisher.PublishAsync(transferRequestUpdated);
+            return transferRequestUpdated;
+        }
+
+        public async Task<ITransferRequest> UpdateTransferStatusAsync(ITransferRequest transfer)
+        {
+            var t = await GetTransferInfoAsync(transfer.TransferId);
+            if (t == null)
+            {
+                return null;
+            }
+            t.TransferStatus = transfer.TransferStatus;
+            t.TransferStatusError = transfer.TransferStatusError;
+
+            var result =  await _transferRepository.SaveAsync(t) ?? BtcTransferRequest.CreateErrorTransferRequest(transfer.MerchantId,
```

Comment:
I believe you are trying to create a response object but why are you using "request" postfix for that?